irohad/pending_txs_storage/impl/pending_txs_storage_impl.cpp

```diff
@@ -64,9 +68,34 @@ PendingTransactionStorageImpl::getPendingTransactions(
   while (account_batches.batches.end() != batch_iterator
          and remaining_space >= batch_iterator->get()->transactions().size()) {
     auto &txs = batch_iterator->get()->transactions();
-    response.transactions.insert(
-        response.transactions.end(), txs.begin(), txs.end());
-    remaining_space -= txs.size();
+    // can be easily extended by adding more conditions if needed
+    std::function<bool(int64_t)> time_pred_f;
+    if (first_tx_time) {
+      time_pred_f = [&first_tx_time](int64_t tx_tms) {
+        return tx_tms > first_tx_time;
+      };
+    } else {
+      time_pred_f = [](auto tx_tms) { return true; };
+    }
+    std::function<bool(int64_t)> time_pred_l;
+    if (last_tx_time) {
+      time_pred_l = [&last_tx_time](int64_t tx_tms) {
+        return tx_tms < last_tx_time;
+      };
+    } else {
+      time_pred_l = [](auto tx_tms) { return true; };
+    }
+    auto time_predicate = [&time_pred_f, &time_pred_l](auto &tx) {
+      auto created_time = tx->createdTime();
+      return time_pred_f(created_time) and time_pred_l(created_time);
+    };
```

Comment:
```c++
while (account_batches.batches.end() != batch_iterator
       && (response.transactions.size()
           + (*batch_iterator)->transactions().size())
           <= page_size) {
  auto const &txs = batch_iterator->get()->transactions();
  std::copy_if(txs.begin(),
               txs.end(),
               std::back_inserter(response.transactions),
               [&first_tx_time, &last_tx_time](auto const &tx) {
                 auto const ts = tx->createdTime();
                 return (!first_tx_time || ts >= *first_tx_time)
                     && (!last_tx_time || ts <= *last_tx_time);
               });
  ++batch_iterator;
}

Reply:
It still not correct. I just retype your loop. But you can see that you compares the whole ```transactions().size()``` with the ```remaining_space```. And uses the part of it ```[first,last]```. So it means you can break the loop even if the required part of txs set can by applied.